File: Plugins/NotesPlugin.cs
using System.ComponentModel;
using Microsoft.SemanticKernel;
using SkOfflineCourse.Infra;

namespace SkOfflineCourse.Plugins;

public class NotesPlugin
{
    private readonly JsonMemoryStore _store;
    private const string Key = "notes";

    public NotesPlugin(JsonMemoryStore store) => _store = store;

    public record Note(string Content, DateTime CreatedAt);

    [KernelFunction, Description("Adiciona uma nota")]
    public async Task<string> AddNote([Description("Conteúdo da nota")] string content)
    {
        var list = await _store.LoadListAsync<Note>(Key);
        list.Add(new Note(content, DateTime.Now));
        await _store.SaveListAsync(Key, list);
        return "📝 Nota salva.";
    }

    [KernelFunction, Description("Lista notas")]
    public async Task<string> ListNotes()
    {
        var list = await _store.LoadListAsync<Note>(Key);
        if (list.Count == 0) return "📭 Sem notas.";
        return string.Join(Environment.NewLine,
            list.Select((n, i) => $"{i+1}. {n.CreatedAt:dd/MM HH:mm} – {Trim(n.Content, 80)}"));
    }

    [KernelFunction, Description("Busca notas por termo (case-insensitive)")]
    public async Task<string> SearchNotes([Description("Termo de busca")] string term)
    {
        var list = await _store.LoadListAsync<Note>(Key);
        var q = term?.Trim() ?? "";
        if (string.IsNullOrWhiteSpace(q)) return "❌ Informe um termo.";
        var found = list.Select((n, i) => (note: n, idx: i+1))
                        .Where(x => x.note.Content.Contains(q, StringComparison.OrdinalIgnoreCase))
                        .ToList();
        if (found.Count == 0) return "🔎 Nenhuma nota encontrada.";
        return string.Join(Environment.NewLine,
            found.Select(x => $"{x.idx}. {Trim(x.note.Content, 80)}"));
    }

    [KernelFunction, Description("Resumo curto e determinístico da nota (sem LLM)")]
    public async Task<string> SummarizeNote([Description("Índice da nota (1-based)")] int index)
    {
        var list = await _store.LoadListAsync<Note>(Key);
        if (index < 1 || index > list.Count) return "❌ Índice inválido.";
        var text = list[index - 1].Content;

        // “Resumo” determinístico: pega 1ª frase ou até 120 chars + “...”
        var dot = text.IndexOf('.', StringComparison.Ordinal);
        string summary = dot > 0 ? text[..(dot+1)] : Trim(text, 120);
        return $"🧾 Resumo: {summary}";
    }

    private static string Trim(string s, int max) =>
        s.Length <= max ? s : s[..max] + "...";
}

File: codigo-python/notes_plugin.py
from dataclasses import dataclass
from datetime import datetime
from typing import List
from json_memory_store import JsonMemoryStore
from summarizer import ISummarizer

@dataclass
class Note:
    content: str
    created_at: str

class NotesPlugin:
    def __init__(self, store: JsonMemoryStore, summarizer: ISummarizer):
        self.store = store
        self.summarizer = summarizer
        self.key = "notes"
    
    async def add_note(self, content: str) -> str:
        notes = await self.store.load_list_async(self.key, dict)
        note_dict = {
            "content": content,
            "created_at": datetime.now().strftime("%d/%m %H:%M")
        }
        notes.append(note_dict)
        await self.store.save_list_async(self.key, notes)
        return "Nota salva."
    
    async def list_notes(self) -> str:
        notes = await self.store.load_list_async(self.key, dict)
        if not notes:
            return "Sem notas."
        
        result = []
        for i, note in enumerate(notes, 1):
            content = self._trim(note.get('content', ''), 80)
            created_at = note.get('created_at', '')
            result.append(f"{i}. {created_at} – {content}")
        
        return "\n".join(result)
    
    async def search_notes(self, term: str) -> str:
        notes = await self.store.load_list_async(self.key, dict)
        if not term or not term.strip():
            return "Informe um termo."
        
        found = []
        for i, note in enumerate(notes, 1):
            if term.lower() in note.get('content', '').lower():
                content = self._trim(note.get('content', ''), 80)
                found.append(f"{i}. {content}")
        
        if not found:
            return "Nenhuma nota encontrada."
        
        return "\n".join(found)
    
    async def summarize_note(self, index: int) -> str:
        notes = await self.store.load_list_async(self.key, dict)
        if index < 1 or index > len(notes):
            return "Índice inválido."
        
        text = notes[index - 1].get('content', '')
        summary = self.summarizer.summarize(text, 120)
        return f"Resumo: {summary}"
    
    @staticmethod
    def _trim(text: str, max_length: int) -> str:
        return text if len(text) <= max_length else text[:max_length] + "..."